### PR TITLE
Laboratory work #4, Lilia Sabirova - 23FPL1

### PR DESCRIPTION
**Pull request by SabirovaLilia**

### CODE REVIEW

File: lab_4_retrieval_w_clustering/main.py [hunk 1]
```diff
@@ -40,6 +46,9 @@ def __init__(self) -> None:
         """
         Initialize an instance of the BM25Vectorizer class.
         """
+        self._corpus = []
+        self._avg_doc_len = -1.0
+        Vectorizer.__init__(self, self._corpus)
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:51]: use `super`

File: lab_4_retrieval_w_clustering/main.py [hunk 2]
```diff
@@ -51,6 +60,14 @@ def set_tokenized_corpus(self, tokenized_corpus: TokenizedCorpus) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not tokenized_corpus or not isinstance(tokenized_corpus, list):
+            raise ValueError
+        self._corpus = tokenized_corpus
+
+        tokens = sum(len(doc) for doc in tokenized_corpus)
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:67]: why do you need this variables?

File: lab_4_retrieval_w_clustering/main.py [hunk 3]
```diff
@@ -67,6 +84,15 @@ def vectorize(self, tokenized_document: list[str]) -> Vector:
...
             Vector: BM25 vector for document.
         """
+        if (tokenized_document is None or
+                not tokenized_document or
+                not isinstance(tokenized_document, list) or
+                not all(isinstance(elem, str) for elem in tokenized_document)):
+            raise ValueError
+        vector_tok_doc = self._calculate_bm25(tokenized_document)
+        if not vector_tok_doc:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:94]: add message to the error

File: lab_4_retrieval_w_clustering/main.py [hunk 4]
```diff
@@ -81,6 +107,21 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
...
             Vector: BM25 vector for document.
         """
+        if not isinstance(tokenized_document, list) or not all(isinstance(elem, str) for elem in tokenized_document):
+            raise ValueError
+
+        calculated_bm25 = [0.0] * len(self._vocabulary)
+
+        idf_document = calculate_idf(self._vocabulary, self._corpus)
+        doc_len = len(tokenized_document)
+        k1 = 1.5
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:117]: do not create this constants

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Good start

File: lab_4_retrieval_w_clustering/main.py (hunk 4 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 1 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/settings.json [hunk 5]
```diff
@@ -1,3 +1,3 @@
 {
-    "target_score": 0
+    "target_score": 6
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/settings.json:2]: What about `start.py`?

File: lab_4_retrieval_w_clustering/main.py [hunk 6]
```diff
@@ -182,6 +259,22 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
...
+                not isinstance(n_neighbours, int) or
+                n_neighbours <= 0):
+            raise ValueError
+
+        tokenized_query = self._tokenizer.tokenize(query)
+        query_vectors = self._db.get_vectorizer().vectorize(tokenized_query)
+        document_vectors = [vector[1] for vector in self._db.get_vectors()]
+        calculated_knn = self._calculate_knn(query_vectors, document_vectors, n_neighbours)
+        if not calculated_knn:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:274]: add message

File: lab_4_retrieval_w_clustering/main.py [hunk 7]
```diff
@@ -182,6 +259,22 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
         Returns:
             list[tuple[float, str]]: Relevant documents with their distances.
         """
+        if (not query or
+                not isinstance(query, str) or
+                not n_neighbours or
+                not isinstance(n_neighbours, int) or
+                n_neighbours <= 0):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:267]: add message

File: lab_4_retrieval_w_clustering/main.py [hunk 8]
```diff
@@ -113,6 +155,23 @@ def put_corpus(self, corpus: Corpus) -> None:
...
+
+        tokenized_docs = []
+        for document in corpus:
+            list_of_tok = self._tokenizer.tokenize(document)
+            if list_of_tok:
+                tokenized_docs.append(list_of_tok)
+                self.__documents.append(document)
+        if not tokenized_docs:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:168]: add message

File: lab_4_retrieval_w_clustering/main.py [hunk 9]
```diff
@@ -113,6 +155,23 @@ def put_corpus(self, corpus: Corpus) -> None:
                 or if input arguments are empty,
                 or if methods used return None.
         """
+        if not corpus or not isinstance(corpus, list) or not all(isinstance(elem, str) for elem in corpus):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:159]: add message

File: lab_4_retrieval_w_clustering/main.py [hunk 10]
```diff
@@ -81,6 +104,21 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if (not tokenized_document or
+                not isinstance(tokenized_document, list) or
+                not all(isinstance(elem, str) for elem in tokenized_document)):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:110]: add message

File: lab_4_retrieval_w_clustering/main.py [hunk 11]
```diff
@@ -26,6 +28,10 @@ def get_paragraphs(text: str) -> list[str]:
     Returns:
         list[str]: Paragraphs from document.
     """
+    if not text or not isinstance(text, str):
+        raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:32]: add message

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 6 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py:274]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 7 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py:267]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 8 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py:168]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 10 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py:110]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 11 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py:32]: done

File: lab_4_retrieval_w_clustering/main.py [hunk 12]
```diff
@@ -4,8 +4,16 @@
 Vector search with clusterization
 """
 
+from lab_2_retrieval_w_bm25.main import calculate_bm25
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: all imports together after pylint disable

File: lab_4_retrieval_w_clustering/main.py [hunk 13]
```diff
@@ -40,6 +52,9 @@ def __init__(self) -> None:
         """
         Initialize an instance of the BM25Vectorizer class.
         """
+        self._corpus = []
+        self._avg_doc_len = -1.0
+        super().__init__(self._corpus)
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: should be the first line in this method

File: lab_4_retrieval_w_clustering/main.py [hunk 14]
```diff
@@ -81,6 +110,21 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
...
+        if (not tokenized_document or
+                not isinstance(tokenized_document, list) or
+                not all(isinstance(elem, str) for elem in tokenized_document)):
+            raise ValueError('Invalid arguments')
+
+        calculated_bm25 = [0.0] * len(self._vocabulary)
+        doc_len = len(tokenized_document)
+
+        bm25 = calculate_bm25(self._vocabulary, tokenized_document,
+                              self._idf_values, 1.5, 0.75, self._avg_doc_len, doc_len)
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: do not set `1.5, 0.75` - it is argument by default

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
What about `start.py`?

File: lab_4_retrieval_w_clustering/main.py (hunk 14 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 13 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 12 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py [hunk 15]
```diff
@@ -4,8 +4,16 @@
 Vector search with clusterization
 """
 
+
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: remove

File: lab_4_retrieval_w_clustering/main.py [hunk 16]
```diff
@@ -182,6 +266,24 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
...
+        if (not query or
+                not isinstance(query, str) or
+                not n_neighbours or
+                not isinstance(n_neighbours, int) or
+                n_neighbours <= 0):
+            raise ValueError('Invalid arguments')
+
+        tokenized_query = self._tokenizer.tokenize(query)
+        if not tokenized_query or not isinstance(tokenized_query, list):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: add message

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 16 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 15 above)
Comment (SabirovaLilia) [reply lab_4_retrieval_w_clustering/main.py]: done

**Pull request closed without merging**